circuit-macros/src/circuit_type.rs

```diff
@@ -0,0 +1,368 @@
+        let multiprover_type_tokens = build_multiprover_circuit_types(&target_struct);
+        out_tokens.extend(multiprover_type_tokens)
+    }
+
+    // Build the commitment-linkable type
+    if macro_args.build_linkable_types {
+        let linkable_type_tokens =
+            build_linkable_types(&target_struct, macro_args.build_multiprover_linkable_types);
+        out_tokens.extend(linkable_type_tokens);
+    }
+
+    // Build secret share types
+    if macro_args.build_secret_share_types {
+        let secret_share_type_tokens = build_secret_share_types(&target_struct);
+        out_tokens.extend(secret_share_type_tokens);
+    }
+
+    out_tokens.into()
+}
+
+// ---------------------------
+// | BaseType Implementation |
+// ---------------------------
+
+/// Build the `impl BaseType` block
+fn build_base_type_impl(base_type: &ItemStruct) -> TokenStream2 {
+    let trait_ident = new_ident(BASE_TYPE_TRAIT_NAME);
+    let base_type_ident = base_type.ident.clone();
+    let scalar_type_path = path_from_ident(new_ident(SCALAR_TYPE_IDENT));
+
+    let from_scalars_impl = build_deserialize_method(
+        Ident::new(FROM_SCALARS_METHOD_NAME, Span::call_site()),
+        scalar_type_path.clone(),
+        path_from_ident(trait_ident.clone()),
+        base_type,
+    );
+
+    let to_scalars_impl = build_serialize_method(
+        Ident::new(TO_SCALARS_METHOD_NAME, Span::call_site()),
+        scalar_type_path,
+        base_type,
+    );
+
+    let impl_block: ItemImpl = parse_quote! {
+        impl #trait_ident for #base_type_ident {
+            #from_scalars_impl
+            #to_scalars_impl
+        }
+    };
+    impl_block.to_token_stream()
+}
+
+// -----------
+// | Helpers |
+// -----------
+
+/// A helper that specifies the default call site span for an Identifier
+fn new_ident(name: &str) -> Ident {
+    Ident::new(name, Span::call_site())
+}
+
+/// A helper that creates an identifier with the given prefix
+fn ident_with_prefix(original: &str, prefix: &str) -> Ident {
+    new_ident(&format!("{prefix}{original}"))
+}
+
+/// A helper to strip a prefix from an identifier and return a new identifier
+fn ident_strip_prefix(original: &str, prefix: &str) -> Ident {
+    let stripped = original.strip_prefix(prefix).unwrap_or(original);
+    new_ident(stripped)
+}
+
+/// A helper that creates an identifier with the given suffix
+fn ident_with_suffix(original: &str, suffix: &str) -> Ident {
+    new_ident(&format!("{original}{suffix}"))
+}
+
+/// A helper to strip a suffix from an identifier and return a new identifier
+fn ident_strip_suffix(original: &str, suffix: &str) -> Ident {
+    let stripped = original.strip_suffix(suffix).unwrap_or(original);
+    new_ident(stripped)
+}
+
+/// Convert a string to a `Path` syntax tree object representing a type path
+fn str_to_path(s: &str) -> Path {
+    path_from_ident(new_ident(s))
+}
+
+/// Convert an `Ident` directly into a `Path`
+fn path_from_ident(identifier: Ident) -> Path {
```

Comment:
would be nifty for this to do full path resolution, but eh, not needed in this PR